File: bench/memcompare/klio/numeric.kt
// Large primitive-array workload: sieve of Eratosthenes up to a bound, then a
// reduction over the first K primes. Stresses one big flat array with little GC
// pressure — the opposite profile from the collections workload.

fun main() {
    val limit = 10_000_000
    val sieve = BooleanArray(limit + 1)
    var count = 0
    var i = 2
    while (i <= limit) {
        if (!sieve[i]) {
            count++
            if (i <= 3163) {
                var j = i * i
                while (j <= limit) { sieve[j] = true; j += i }
            }
        }
        i++
    }
    val k = 100_000
    var taken = 0
    var acc = 0L
    var p = 2
    while (p <= limit && taken < k) {
        if (!sieve[p]) {
            acc = (acc + p.toLong() * p) % 1_000_000_007L
            taken++
        }
        p++
    }
    println("numeric: limit=$limit primes=$count firstK=$taken checksum=$acc")
}

File: bench/memcompare/node/numeric.js
// Large primitive-array workload — see klio/numeric.kt for the shared spec.
function main() {
  const limit = 10000000;
  const sieve = new Uint8Array(limit + 1);
  let count = 0;
  for (let i = 2; i <= limit; i++) {
    if (sieve[i] === 0) {
      count++;
      if (i <= 3163) { for (let j = i * i; j <= limit; j += i) sieve[j] = 1; }
    }
  }
  const k = 100000;
  let taken = 0, acc = 0n;
  for (let p = 2; p <= limit && taken < k; p++) {
    if (sieve[p] === 0) { acc = (acc + BigInt(p) * BigInt(p)) % 1000000007n; taken++; }
  }
  console.log(`numeric: limit=${limit} primes=${count} firstK=${taken} checksum=${acc}`);
}
main();
